feat(surveys): add branching logic

jurajmajerik opened the pull request:
## Changes
^^

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [x] Accounted for the impact of any changes across different browsers
- [x] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

jurajmajerik commented on src/extensions/surveys/surveys-utils.tsx:
```diff
@@ -671,3 +678,105 @@ export const renderChildrenAsTextOrHtml = ({ component, children, renderAsHtml,
+    } else if (question.branching.type === SurveyQuestionBranchingType.ResponseBased) {
+        // Single choice
+        if (question.type === SurveyQuestionType.SingleChoice) {
+            // :KLUDGE: for now, look up the choiceIndex based on the response
+            // TODO: once QuestionTypes.MultipleChoiceQuestion is refactored, pass the selected choiceIndex into this method
```
Will address this kludge later to not interfere with @dmarticus refactor

neilkakkar commented on src/extensions/surveys/surveys-utils.tsx:
```diff
@@ -671,3 +678,105 @@ export const renderChildrenAsTextOrHtml = ({ component, children, renderAsHtml,
+            // TODO: once QuestionTypes.MultipleChoiceQuestion is refactored, pass the selected choiceIndex into this method
+            const selectedChoiceIndex = question.choices.indexOf(`${response}`)
+
+            if (question.branching?.responseValues?.hasOwnProperty(selectedChoiceIndex)) {
```
We should make this part of the SDK like `getActiveMatchingSurveys`, to `getNextQuestion(survey, currentQuestionIndex, response) -> nextQuestionIndex` for better API mode handling.

Plus, this means we can much more flexibly change things here later, which is bound to happen imo (new branching options, changing brackets, etc. etc.) - all of which would be very hard to handle for API mode if we don't have it as an SDK level primitive.

jurajmajerik replied:
makes sense and done - named it `getNextSurveyStep` since it returns either `index` or `confirmation_message`

neilkakkar commented on src/extensions/surveys.tsx:
```diff
@@ -356,16 +363,20 @@ export function Questions({
         displayQuestionIndex: number
     }) => {
-        const isLastDisplayedQuestion = displayQuestionIndex === survey.questions.length - 1
+        if (!posthog) {
```
when does this happen?

jurajmajerik replied:
`posthog` is an optional argument to `SurveyPopup` (we're not passing it in the preview mode), so this needs to be checked

neilkakkar commented on src/extensions/surveys.tsx:
```diff
@@ -356,16 +363,20 @@ export function Questions({
-            return sendSurveyEvent({ ...questionsResponses, [responseKey]: res }, survey, posthog)
+        setQuestionsResponses({ ...questionsResponses, [responseKey]: res })
+
+        const nextStep = posthog?.getNextSurveyStep(survey, displayQuestionIndex, res)
```
Here, `nextStep` should always exist right, so 
```suggestion
        const nextStep = posthog.getNextSurveyStep(survey, displayQuestionIndex, res)
```

I'd expect the above to work^ if it doesn't there's potentially a bug? (or ts being dumb)

neilkakkar commented on src/posthog-surveys.ts:
```diff
@@ -155,4 +161,101 @@ export class PostHogSurveys {
+            // Single choice
+            if (question.type === SurveyQuestionType.SingleChoice) {
+                // :KLUDGE: for now, look up the choiceIndex based on the response
+                // TODO: once QuestionTypes.MultipleChoiceQuestion is refactored, pass the selected choiceIndex into this method
```
I think now this is independent of any refactors?

Reminder that choices can be shuffled too, would be nice to see some tests for the single choice case + where the choices are shuffled

jurajmajerik replied:
Should've been clearer - I want to refactor `MultipleChoiceQuestion ` into `SingleSelectQuestion` and `MultipleSelectQuestion`. When that's done, I'll pass `choiceIndex` into this method directly rather than looking it up with the choice value. Want to do it later to not interfere with Dylan (who's also working on this file).

Added the test for single choice w/ shuffling 👍

neilkakkar commented on src/posthog-surveys.ts:
```diff
@@ -155,4 +161,101 @@ export class PostHogSurveys {
+                    }
+                }
+
+                const resNumber = Number(response)
```
this needs to defend against invalid input, since users can mistakenly input `choiceX` in response - we should throw a more appropriate error.

neilkakkar commented on src/posthog-surveys.ts:
```diff
@@ -155,4 +161,101 @@ export class PostHogSurveys {
+                }
+
+                const resNumber = Number(response)
+                const ratingBucket = responseValueToBucket[resNumber]
```
minor nit: The object is hard to parse, specially because of the repeated strings. Since it's only used once, I think works better as a function `getBucketForResponseValue` that just compares the no. to return the bucket.

return resNumber <= 6 ? 'detractors' : resNumber <= 8 ? 'passives' : 'promoters'

neilkakkar commented on src/posthog-surveys.ts:
```diff
@@ -155,4 +161,101 @@ export class PostHogSurveys {
+                }
+            }
+
+            return nextQuestionIndex
```
I think `console.warn`s whenever we fallback because of invalid input would be useful - there's no case I can see where if things are setup correctly we reach here

neilkakkar replied:
Looks good! Are there any other cases where it's an unexpected return, or are they all for the implicit next question case?

neilkakkar reviewed (commented):
Nicely done!

jurajmajerik commented:
Changes:
- added a test for single-choice + shuffling
- defend against non-integer input when looking up response bucket
- added getRatingBucketForResponseValue method
- added console.warn when unexpected branching type

jurajmajerik commented on src/__tests__/surveys.test.ts:
```diff
@@ -782,4 +789,301 @@ describe('surveys', () => {
+            end_date: null,
+        } as Survey
+
+        // Simple branching
```
tests added :) https://github.com/PostHog/posthog-js/pull/1247/commits/3d0e32b79ee75e15cf068de72bc3b1837bc81909
